ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerStateActions.hs

```diff
@@ -624,15 +622,30 @@ withPeerStateActions PeerStateActionsArguments {
         -- this is a first-to-finish synchronisation between all the
         -- mini-protocols runs toward the given peer.
         r <-
-          atomically $
-            (WithSomeProtocolTemperature . WithEstablished
-              <$> awaitFirstResult SingEstablished pchAppHandles)
-          `orElse`
-            (WithSomeProtocolTemperature . WithWarm
-              <$> awaitFirstResult SingWarm pchAppHandles)
-          `orElse`
-            (WithSomeProtocolTemperature . WithHot
-              <$> awaitFirstResult SingHot pchAppHandles)
+          atomically $ do
+            peerStatus <- readTVar pchPeerStatus
+            -- If we already monitored mux and are waiting for the peer connection
+            -- to be cleaned then the peer status will be 'PeerCold' and we can't
+            -- make progress until it is 'PeerReallyCold'
+            case peerStatus of
+              PeerCold ->
+                return Nothing
+              PeerCooling -> do
+                cmState <- readState spsConnectionManager
+                case Map.lookup (remoteAddress pchConnectionId) cmState of
+                  Just UnknownConnectionSt -> do
+                    writeTVar pchPeerStatus PeerCold
+                    return Nothing
+                  _ -> retry
```

Comment:
I think this can lock the peer in a PeerCooling state. Consider a duplex connection where the outbound side terminates, the peer becomes PeerCooling until connection manager will forget the connection, so until the inbound side terminates.